File: Functions3/Game Logic/GameLogic.m
//
//  GameLogic.m
//  Functions3
//
//  Created by erdikanik on 30.08.2015.
//  Copyright (c) 2015 ekanik. All rights reserved.
//

#import "GameLogic.h"
#import "FMath.h"
#import "FunctionNumber.h"
#import "Functions.h"

const CGFloat kGameLogicWaitForDuration = 0.01;
const CGFloat kGameLogicRandomFunctionsDuration = 10;

@interface GameLogic()

@property (assign, nonatomic) NSTimeInterval initialTime;
@property (assign, nonatomic) NSTimeInterval scoreTime;
@property (strong, nonatomic) NSTimer *timer;
@property (strong, nonatomic) NSTimer *methodTimer;
@property (assign, nonatomic) double totalNumber;
@property (strong, nonatomic) Polynomal *function;

@end

@implementation GameLogic

- (instancetype)initWithGameOverTime:(NSTimeInterval)initialTime
{
    self = [super init];
    if (self)
    {
        _initialTime = initialTime;
        _scoreTime = initialTime;
    }

    return self;
}

- (void)gameStarted {

    self.timer = [NSTimer scheduledTimerWithTimeInterval:kGameLogicWaitForDuration repeats:YES block:^(NSTimer * _Nonnull timer) {
       if (self.scoreTime <= 0)
       {
           [self.timer invalidate];
           [self.delegate timeUpdated:self time:self.scoreTime gameOvered:YES];
       }
       else
       {
           self.scoreTime -= kGameLogicWaitForDuration;
           [self.delegate timeUpdated:self time:self.scoreTime gameOvered:NO];
       }
    }];

    [self initializeRandomFunctions];
}

- (void)setNumber:(double)number
{
    _number = number;
    _totalNumber = [self.function getPolynomalValue:_number];
    self.score += self.totalNumber > 0 ? self.totalNumber * 0.1 : 0;
    _scoreTime += _totalNumber * 0.5;
    [self.delegate functionResulted:self functionResult:_totalNumber];
}

#pragma mark - Functions

- (void)initializeRandomFunctions
{
    self.timer = [NSTimer scheduledTimerWithTimeInterval:kGameLogicRandomFunctionsDuration
                                                 repeats:YES
                                                   block:^(NSTimer * _Nonnull timer) {
        [self getRandomFunction];
    }];
}

- (void)getRandomFunction
{
    self.function = [Functions getRandomFunction];
    [self.delegate functionChanged:self function:self.function];
}

+ (FNumber*)getNumberFromLogic
{
    CGFloat num = [FMath getRandomGeneratorGivenRange:0 withSecond:1000];
    
    if (num < 800)
    {
        FNumber *number = [[FNumber alloc] initWithNumber:[self getRandomNumber]];
        return number;
    }
    else if (num < 850)
    {
        FNumber *number = [[FNumber alloc] initWithType:FNumberTypeEmerald];
        return number;
    }
    else if (num < 900)
    {
        FNumber *number = [[FNumber alloc] initWithType:FNumberTypeDiamond];
        return number;
    }
    else if (num < 950)
    {
        FNumber *number = [[FNumber alloc] initWithType:FNumberTypeGolden];
        return number;
    }
    else if (num < 850)
    {
        FNumber *number = [[FNumber alloc] initBombTypeWithNumber:[self getRandomNumber]];
        return number;
    }
    else if (num < 900)
    {
        CGFloat num1 = [GameLogic getRandomNumber];
        CGFloat num2 = [GameLogic getRandomNumber];
        CGFloat num3 = [GameLogic getRandomNumber];
        NSArray *numberArray = [NSArray arrayWithObjects:_N(num1),_N(num2),_N(num3),nil];
        FNumber *number = [[FNumber alloc] initWithNumberArray:numberArray];
        return number;
    }
    else if (num < 940)
    {
        FNumber *number = [[FNumber alloc] initBombTypeWithNumber:[self getRandomNumber]];
        return number;
    }
    else
    {
        FNumber *number = [[FNumber alloc] initInvisibleNumber:[self getRandomNumber]];
        return number;
    }
    
}

+ (CGFloat)getRandomNumber
{
    return [FMath getRandomGeneratorGivenRange:-10 withSecond:9];
}

@end

File: Functions3/Game Logic/GameLogic.swift
//
//  GameLogic.swift
//  Functions3
//
//  Created by Erdi Kanik on 11.11.2017.
//  Copyright © 2017 ekanik. All rights reserved.
//

import Foundation

fileprivate enum Constants {
    static let gameLogicWaitDuration: TimeInterval = 0.01
    static let gameLogicRandomFunctionsDuration: Float = 500
}

@objc protocol GameLogicDelegate: class {
    func gameLogic(_ gameLogic: GameLogic, time:TimeInterval, gameOvered: Bool)
    func gameLogic(_ gameLogic: GameLogic, functionResulted result: Int)
    func gameLogic(_ gameLogic: GameLogic, functionChanged function: ExactMethod)
    func gameLogic(_ gameLogic: GameLogic, functionTimeDidChanged time: Float)
    func gameLogic(_ gameLogic: GameLogic, levelChanged level: Int)
}

@objc class GameLogic: NSObject {
    weak var delegate: GameLogicDelegate?
    @objc var number: Int = 0 {
        didSet {
            if let num = polynomal?.getValue(CGFloat(number)) {
                let result = Double(num)
                score += result > 0 ? result * 0.05 : 0
                scoreTime += result > 0 ? result * 0.2 : result + (Double(self.level) * 0.01)
                delegate?.gameLogic(self, functionResulted: Int(result))
            }
        }
    }

    var score: Double = 0

    private var scoreTime: TimeInterval = 1
    private var timer: Timer?
    private var functionTimer: Timer?
    private var polynomal: Polynomal?
    private var exactMethod: ExactMethod?
    private var functionChangingTime: Float = 0
    private var level: Int = 1
    private var initialTime: Int = 100
    private var latestResult = 0

    required init(gameOverTime: TimeInterval) {
        super.init()
        self.scoreTime = gameOverTime
    }

    func gameStarted() {
        randomFunction()
        self.functionChangingTime = Constants.gameLogicRandomFunctionsDuration
        self.timer = Timer.scheduledTimer(withTimeInterval: Constants.gameLogicWaitDuration, repeats: true) {
            [weak self] (timer) in

            guard let strongSelf = self else {
                return
            }

            if strongSelf.scoreTime <= 0 {
                strongSelf.timer?.invalidate()
                strongSelf.delegate?.gameLogic(strongSelf, time: strongSelf.scoreTime, gameOvered: true)
            } else {
                strongSelf.scoreTime -= Constants.gameLogicWaitDuration
                strongSelf.delegate?.gameLogic(strongSelf, time: strongSelf.scoreTime, gameOvered: false)
            }

            strongSelf.delegate?.gameLogic(strongSelf, functionTimeDidChanged: strongSelf.functionChangingTime / 10)
            strongSelf.functionChangingTime -= 1

            if strongSelf.functionChangingTime <= 0 {
                strongSelf.randomFunction()
                strongSelf.functionChangingTime = Constants.gameLogicRandomFunctionsDuration
            }

            strongSelf.initialTime += 1
            strongSelf.levelDetection()
        }
    }

    private func randomFunction() {
        exactMethod = ExactMethod()
        delegate?.gameLogic(self, functionChanged: exactMethod!)
    }

    private func levelDetection() {

        let doubleLevel = Double(level)
        let formula = 100 * doubleLevel * (level > 20 ? doubleLevel * (doubleLevel * 0.1) : 5)
        if (initialTime / Int(formula)) > 0 {
            level += 1
            delegate?.gameLogic(self, levelChanged: level)
        }
    }
}
